### ISSUE TITLE
Feature Request: Debug run

### ISSUE THREAD

**Issue by romangraef** [EntryPointKR/Spigradle#5]
Are there plans for this plugin to support a debug run feature? This could be implemented using an extra task. If there is no such thing planned, i might even do it myself, if other interest exists.

**Comment by EntryPointKR** [comment]
Can you please explain what exactly the "debug run" do?
There are no such debug feature and no plans now, but if is it helpful, I will consider to add it.

**Comment by romangraef** [comment]
Sorry if i was being unclear. I meant something like a in-IDE run, or a server run using the gradle project. Using forge as an example, there is a gradle task that start a modded client with your mod installed. I was thinking of something similar by simply downloading buildtools and creating a server instance with your plugin installed automatically.

**Comment by EntryPointKR** [comment]
Thanks for the explain :)

**Comment by NickAcPT** [comment]
Would also love to see this! It would make a lot of development simple.

For CraftBukkit and Spigot, we'll need to run buildtools. For Paper, they have a web API where you can use to directly download a Paperclip jar that automatically patches the vanilla jar.

**Comment by romangraef** [comment]
I almost forgot about this. I can definitely provide some working code for it, however it is in form of a buildSrc/ folder right now, and i have a few too many comissions to do normal open source work and convert it into a plugin extension. If anyone is interested i can do that, but i think we should think about the actual structure of this inbefore.

**Comment by NickAcPT** [comment]
I'd maybe suggest adding a task named `runServer`. And, when started, it'll make a `run` folder on the project folder (like ForgeGradle), if it not exists.
That folder will then have the server files.
It would be nice to have a way to add other plugins (like dependencies or something else).
Also, if possible, make it pass `nogui` or let us pass arbitrary arguments to the task.

**Comment by romangraef** [comment]
Yes, that + as project extension for easy configuration of the server. (Like maybe an auto eula accept option, or command arguments). However I'm not quite sure how you make project extensions available using gradle projects.

**Comment by NickAcPT** [comment]
I think FG allows the users to configure stuff directly from Gradle.
Would be nice to make it so anyone can just `git clone` a project and start working on it and having it ready to run.

**Comment by portlek** [comment]
breakpoints can be useful when someone who write plugins.

**Comment by romangraef** [comment]
i dont think breakpoints can be easily done without bootstrapping the server directly (which might be possible, but would be very hard)

**Comment by portlek** [comment]
its possible but before that this debug stuffs should finish. breakpoints are depend debug tools.

**Comment by NickAcPT** [comment]
> i dont think breakpoints can be easily done without bootstrapping the server directly (which might be possible, but would be very hard)

Don't think you need to complicate everything. Running the server JAR inside Intellij already works wonders. If the code is loaded, the debugger should set a breakpoint automatically.

**Comment by romangraef** [comment]
Oh, right, you can run the jar directly in intellij.

**Comment by portlek** [comment]
The main reason to should be that is not just running debug mode, also running a real sever by C.I. This feature will be helpful in this case too.

**Comment by portlek** [comment]
I added this feature into my scope. I will try to make it.

**Comment by TheAifam5** [comment]
There u can get some inspirations:
- https://gitlab.com/Firestar99/spigotgradle
- https://github.com/EndlessCodeGroup/BukkitGradle

:)

**Comment by NickAcPT** [comment]
Although the things posted by @TheAifam5 are a start, please don't make Spigradle download Spigot builds off wonky of websites. The best is to download BuildTools and build it manually as builds on those websites are usually outdated at best, and dubious at worse.

For Paper, there's their Paperclip tool which automatically downloads and patches a vanilla jar, so you can run it pretty easily. I know that they have somewhere an API for https://papermc.io/downloads

**Comment by EntryPointKR** [comment]
Thanks for all the nice discussion, yeah I'll use BuildTools for build spigot also paper.
Recently I completed migration to kotlin, support nukkit and bungeecord. so now priority is debug run things.
I'll consider to implement passing parameters like nogui, accept eula, download dependency plugins. And, good to hear that you'll make breakpoint feature @portlek.

**Comment by NickAcPT** [comment]
@EntryPointKR I'm happy to see improvements being done as time goes on. 
> yeah I'll use BuildTools for build spigot also paper.

I'm not sure if in here you meant that you were gonna use Build Tools to build Spigot *and* Paper. I'm asking because just wanted to make it clear that only Spigot requires Build Tools action.

> And, good to hear that you'll make breakpoint feature
Breakpoints are supposedly done by the debugger and shouldn't need extra code.

**Comment by EntryPointKR** [comment]
@NickAcPT Good point! I didn't research about paper build yet, so I thought that need to build paper using paper's build tool. but it can be done by paperclip.
And I something misinterpreted. as portlek said breakpoint, I thought it is some special feature because I saw he's code that generates IDEA xml on he's repository.
Before implement these things, I really need to research about it.

**Comment by EntryPointKR** [comment]
But I'm sharing about its importance. I will implement it carefully.

**Comment by TheAifam5** [comment]
@NickAcPT wait what? They use the official build jar file ;D

**Comment by NickAcPT** [comment]
> @NickAcPT wait what? They use the official build jar file ;D

I feel like you're talking about getbukkit. It's usually not a good way to get your JARs. As I've mentioned previously, they're usually pretty much outdated.

**Comment by portlek** [comment]
> > @NickAcPT wait what? They use the official build jar file ;D
> 
> @TheAifam5 I feel like you're talking about getbukkit. It's usually not a good way to get your JARs. As I've mentioned previously, they're usually pretty much outdated.

https://github.com/EndlessCodeGroup/BukkitGradle
this plugin is using original BuildTool.jar file.

**Comment by TheAifam5** [comment]
@NickAcPT I know what I am talking about:

https://github.com/EndlessCodeGroup/BukkitGradle/blob/711bd72fe9a04564d66ab747ddfa4dcfbb726ba9/src/main/groovy/ru/endlesscode/bukkitgradle/server/ServerCore.groovy#L104

https://gitlab.com/Firestar99/spigotgradle/-/blob/master/src/main/java/firestar99/spigotgradle/server/SpigotServerPlugin.java#L28

**Comment by NickAcPT** [comment]
> @NickAcPT I know what I am talking about:
> 
> [EndlessCodeGroup/BukkitGradle:src/main/groovy/ru/endlesscode/bukkitgradle/server/ServerCore.groovy@`711bd72`#L104](https://github.com/EndlessCodeGroup/BukkitGradle/blob/711bd72fe9a04564d66ab747ddfa4dcfbb726ba9/src/main/groovy/ru/endlesscode/bukkitgradle/server/ServerCore.groovy#L104)
> 
> [gitlab.com/Firestar99/spigotgradle/-/blob/master/src/main/java/firestar99/spigotgradle/server/SpigotServerPlugin.java#L28](https://gitlab.com/Firestar99/spigotgradle/-/blob/master/src/main/java/firestar99/spigotgradle/server/SpigotServerPlugin.java#L28)

@TheAifam5 That is BuildTools and I know.. But I was talking about when you mentioned getbukkit and telling you that you shouldn't use it.

**Comment by TheAifam5** [comment]
@NickAcPT that's why I removed it. It was my attempt to build but then I realize that's bad idea and was throwing the warning that is outdated.

**Comment by EntryPointKR** [comment]
I pushed up the prototype of tasks that BuildTools and RunServer just now. (spigot only in this time)
Demo is here: https://github.com/EntryPointKR/Spigradle/blob/bc673db7fdd53dc3996530b1d4d70424e8655d02/src/test/kotlin/kr/entree/spigradle/spigot/PrepareServerTest.kt
Pull request: #9 
It is not completed yet, still in progress of development. if you find things that some weird or improvements in there being progress, feel free to share here.

And, tasks that download BuildTools, build spigot, and run it is not hard to implement.
But I'm not sure what is the best way to download depend plugins for setup a debug run server.
The option that easy to think is just download it from spigotmc or maven repository.

**Comment by NickAcPT** [comment]
> But I'm not sure what is the best way to download depend plugins for setup a debug run server.

I'd say to either allow the user to use something like Spiget and download through Spigot, or have them manually put the jar file.

**Comment by EntryPointKR** [comment]
Yep, I'll allow manual placing the jars.

**Comment by EntryPointKR** [comment]
Okay, the debug tasks seems done.
Now you guys can try it out by cloning this project: [SpigradleSample](https://github.com/EntryPointKR/SpigradleSample).
Execute task `debugPaper` or `debugSpigot` will startup the server with plugins.
Command: `./gradlew debugPaper` or `./gradlew debugSpigot`

List of debug tasks: `downloadSpigotBuildTools, cleanSpigotBuild, buildSpigot, prepareSpigot, runSpigot, prepareSpigotPlugin, debugSpigot, downloadPaper, debugPaper`

If you import it using IntelliJ IDEA, then the "debug" directory will be marked as excluded, Debug [run configurations](https://www.jetbrains.com/help/idea/run-debug-configuration.html) will be generated. Currently supports IntelliJ IDEA only.

I'll document all of changes and release 1.3.0 in this week. If you have a feedback or improvements then feel free to share here :)

**Comment by NickAcPT** [comment]
Instead of "debug", wouldn't it be less confusing if it was named "run"?

**Comment by EntryPointKR** [comment]
@NickAcPT The `runSpgiot` already used as subtask of `debugSpigot`.
More further, the `debugSpigot` is just lifecycle task(no function) that depends on tasks `downloadBuildTools`, `preparePlugins`, `runSpigot`. the reason of division into subtask is for flexibility, it will be documented in wiki soon. I agree about confusing between run and debug. it can be renamed.

**Comment by TheAifam5** [comment]
Do we need “debug” task if run is already there?

**Comment by EntryPointKR** [comment]
The `runSpigot` task just runs the jar at specified path with configured arguments, and accepts EULA if user agreed. 
But the `debugSpigot` nothing to do itself, but executes tasks `preparePlugin`, `prepareSpigot`, `runSpigot`. That means:

`preparePlugin`([Copy](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Copy.html)): It will copy the project artifact jar and dependency plugins(like ProtocolLib). the depend plugins will automatically resolved from classpath. Related: [DebugTasks.kt](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/common/DebugTask.kt#L48)
`prepareSpigot`([Copy](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Copy.html)): It will download the BuildTools from configured URL, and build the Spigot. and copy the Spigot from build output into configured server directory. Related: [SpigotPlugin.kt#76](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/spigot/SpigotPlugin.kt#L76), [SpigotDebugTask.kt](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/spigot/SpigotDebugTask.kt#L68)
`runSpigot`([JavaExec](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/JavaExec.html)): It will runs the Spigot at configured path with configured (vm)arguments, and accepts EULA if user agreed. Related: [SpigotDebugTask.kt#80](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/spigot/SpigotDebugTask.kt#L80), [DebugTask.kt#34](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/common/DebugTask.kt#L34)

**_debugSpigot_**([Task](https://docs.gradle.org/current/javadoc/org/gradle/api/Task.html)): It is nothing to do, but just executes `preparePlugin`, `prepareSpigot`, `runSpigot`.

So if you delete the debug(server) directory and execute only `runSpigot` then you will get error but if you configure valid jar path at the task, it will work. and I thought the task division would be configurable, reusable. but if there a better idea to constructing these tasks, I'll make the change as open mind.

**Comment by EntryPointKR** [comment]
I thought the runSpigot can be used some special case. but if it not worth,  I'll consider to merge two task run and debug.

**Comment by TheAifam5** [comment]
I think so. Using local jar could be set as a flag or a path to the jar, otherwise would execute the debug tasks you mentioned.

**Comment by EntryPointKR** [comment]
I think more about that, the use case is
1. Any task can [depend on](https://docs.gradle.org/current/javadoc/org/gradle/api/Task.html#dependsOn-java.lang.Object...-) runSpigot, not debugSpigot.
2. Run the server without any preparing tasks, prepareSpigot and preparePlugins.

The first one is useless. second, if a user want to avoid prepare tasks, the user can just [disable](https://docs.gradle.org/current/javadoc/org/gradle/api/Task.html#setEnabled-boolean-) the task or set the [skipWhenExists](https://github.com/EntryPointKR/Spigradle/blob/buildtools/src/main/kotlin/kr/entree/spigradle/module/common/Download.kt#L29) property to true for manual placing server jar.

And, if set a different path in run task, then in debug task will same result. run task is a subset of debug task.

So I think both tasks can be merged. if there a no objections, I'll merge both.

**Comment by EntryPointKR** [comment]
Sorry for being too late, the 1.3.0 released just now.
You can find the migration tips at [Spigot Migration Tips](https://github.com/EntryPointKR/Spigradle/blob/master/docs/spigot_plugin.md#migration-tips).
If you had a problem with it, please open a new issue.

Still the usage is:
```groovy
plugins {
    id 'kr.entree.spigradle' version '1.3.0'
}
```

See also:
[The Spigot plugin](https://github.com/EntryPointKR/Spigradle/blob/master/docs/spigot_plugin.md)
[The Bungeecord plugin](https://github.com/EntryPointKR/Spigradle/blob/master/docs/bungeecord_plugin.md)
[The Nukkit plugin](https://github.com/EntryPointKR/Spigradle/blob/master/docs/nukkit_plugin.md)
[Spigradle Sample](https://github.com/EntryPointKR/SpigradleSample/tree/master/spigot)
[Changelogs](https://github.com/EntryPointKR/Spigradle/releases/tag/v1.3.0)

Thanks again for all the contributions!
